clippy_lints/src/let_underscore.rs

```diff
@@ -58,7 +58,40 @@ declare_clippy_lint! {
     "non-binding let on a synchronization lock"
 }
 
-declare_lint_pass!(LetUnderscore => [LET_UNDERSCORE_MUST_USE, LET_UNDERSCORE_LOCK]);
+declare_clippy_lint! {
+    /// **What it does:** Checks for `let _ = <expr>`
+    /// where expr has a type that implements `Drop`
+    ///
+    /// **Why is this bad?** This statement immediately drops the initializer
+    /// expression instead of extending its lifetime to the end of the scope, which
+    /// is often not intended. To extend the expression's lifetime to the end of the
+    /// scope, use an underscore-prefixed name instead (i.e. _var). If you want to
+    /// explicitly drop the expression, `std::mem::drop` conveys your intention
+    /// better and is less error-prone.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    ///
+    /// Bad:
+    /// ```rust,ignore
+    /// struct Droppable;
+    /// impl Drop for Droppable {
+    ///     fn drop(&mut self) {}
+    /// }
+    /// let _ = Droppable;
+    /// ```
+    ///
+    /// Good:
+    /// ```rust,ignore
+    /// let _droppable = Droppable;
+    /// ```
+    pub LET_UNDERSCORE_DROP,
+    correctness,
```

Comment:
I think this lint should be allow-by-default, since it is too general (e.g. triggers for every `Vec`) for having it deny-by-default. So

```suggestion
    pedantic,
```